### PR TITLE
minor optimization: minor optimization to setRoute to improve the code quality

### PR DESCRIPTION
**Pull request by wbpcode**
Commit Message: minor optimization: minor optimization to improve the code quality
Additional Description:

As title.

Risk Level: Low.
Testing: n/a.
Docs Changes: n/a.
Release Notes: n/a.
Platform Specific Features: n/a.

### CODE REVIEW

**Review by mattklein123 — COMMENTED** [review]
Can you explain more what this accomplishes perf or functionality wise? It's not clear. It's scary code so it doesn't strike me as worth it to change it unless there is a good reason. Thank you.

/wait-any

**Comment by wbpcode** [comment]
I am pretty sure this PR is safe. (but i wont bet my life on it, haha. 🤣)

Seriously speaking, this PR will not bring much performance improvement or new features. It just make things a little better, only very minor little.
I still think it make sense. Even the code base is very complex and scary, we can split the PR to make it easy to review or verify. And we can make things better gradually. 😺 

some comments are added to explain what this PR does extactly.

File: source/common/http/conn_manager_impl.cc [hunk 1]
```diff
@@ -143,15 +145,15 @@ void ConnectionManagerImpl::initializeReadFilterCallbacks(Network::ReadFilterCal
   }
 
   if (config_.idleTimeout()) {
-    connection_idle_timer_ = read_callbacks_->connection().dispatcher().createScaledTimer(
-        Event::ScaledTimerType::HttpDownstreamIdleConnectionTimeout,
-        [this]() -> void { onIdleTimeout(); });
+    connection_idle_timer_ =
+        dispatcher_->createScaledTimer(Event::ScaledTimerType::HttpDownstreamIdleConnectionTimeout,
+                                       [this]() -> void { onIdleTimeout(); });
```
Comment (wbpcode) [inline source/common/http/conn_manager_impl.cc:150]: A very simple replacement from `read_callbacks_->connection().dispatcher()` to single `dispatcher_`. In this class, there are 19 positions need use reference of dispatcher and `read_callbacks_->connection().dispatcher()` is called repeatly in 19 positions.

File: source/common/http/conn_manager_impl.cc [hunk 2]
```diff
@@ -1926,19 +1921,23 @@ void ConnectionManagerImpl::ActiveStream::setRoute(Router::RouteConstSharedPtr r
...
+    cached_cluster_info_ = (nullptr == cluster) ? nullptr : cluster->info();
   }
 
+  // Update the cached route.
+  setCachedRoute({route});
+
+  // Update route and cluster info in the filter manager's stream info.
+  filter_manager_.streamInfo().route_ = std::move(route); // We can move route here.
   filter_manager_.streamInfo().setUpstreamClusterInfo(cached_cluster_info_.value());
+
```
Comment (wbpcode) [inline source/common/http/conn_manager_impl.cc:1940]: This is a more meaningful change. In previous code, We will set route the cached route and stream info. 

Then we will call the `filter_manager_.streamInfo().route()` to access the new route repeatly. Every call will create a temporary copy of shared pointer. (not sure if the compiler will optimize it or not)

The new code re-organize the logic to make it easy to read/understand (this point is more important) and eliminate all unnecessary copies.

File: source/common/http/conn_manager_impl.cc (hunk 1 above)
Comment (mattklein123) [reply source/common/http/conn_manager_impl.cc:150]: Can you split this out into a different PR? There is some memory overhead from this change. Not sure how that compares to (maybe) some CPU savings depending on compiler optimizations.

/wait

Comment (wbpcode) [reply source/common/http/conn_manager_impl.cc:150]: got it.

**Comment by wbpcode** [comment]
sorry for the force push. :(

**Review by mattklein123 — APPROVED** [review]

**Review by mattklein123 — APPROVED** [review]

**Pull request merged**